src/gsumm.c

```diff
@@ -909,16 +909,21 @@ SEXP gmedian(SEXP x, SEXP narmArg) {
   return ans;
 }
 
-static SEXP gfirstlast(SEXP x, const bool first, const int w, const bool headw) {
-  // w: which item (1 other than for gnthvalue when could be >1)
-  // headw: select 1:w of each group when first=true, and (n-w+1):n when first=false (i.e. tail)
+static SEXP gfirstlast(const SEXP x, const bool first, const SEXP nArg, const bool nthvalue, const SEXP narmArg) {
+  if (!IS_TRUE_OR_FALSE(narmArg))
+    error(_("%s must be TRUE or FALSE"), "na.rm");  // # nocov
+  const bool narm = LOGICAL(narmArg)[0];
+  if (!isInteger(nArg) || LENGTH(nArg)!=1 || INTEGER(nArg)[0]<0)
```

Comment:
edgy case at `n=0`

```r
DT = data.table(grp=c(1,1,2,2), a=3:6, b=c(7,NA,NA,NA))
DT[, last(.SD, na.rm=TRUE, n=0), by=grp, verbose=TRUE]
# Argument 'by' after substitute: grp
# Finding groups using forderv ... forder.c received 4 rows and 1 columns
# 0.000s elapsed (0.000s cpu) 
# Finding group sizes from the positions (can be avoided to save RAM) ... 0.000s elapsed (0.000s cpu) 
# lapply optimization changed j from 'last(.SD, na.rm = TRUE, n = 0)' to 'list(last(a, na.rm = TRUE, n = 0), last(b, na.rm = TRUE, n = 0))'
# GForce optimized j to 'list(glast(a, na.rm = TRUE, n = 0), glast(b, na.rm = TRUE, n = 0))'
# Making each group and running j (GForce TRUE) ... gforce initial population of grp took 0.000
# gforce assign high and low took 0.000
# gforce eval took 0.000
# 0.000s elapsed (0.001s cpu) 
#      grp          a             b
#    <num>      <int>         <num>
# 1:     1 1730289304 4.684518e-310
# 2:     2      22076 4.684602e-310
```

Reply:
Good spot. Fixed and more tests in cb8f2010f4bdd1a0547663adc1bac92ee12e823b.